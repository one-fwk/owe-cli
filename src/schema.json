{
  "additionalProperties": false,
  "required": ["projects", "browserTargets"],
  "properties": {
    "browserTargets": {
      "type": "array"
    },
    "projects": {
      "type": "object"
    }
  }
}