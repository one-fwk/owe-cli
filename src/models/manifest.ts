export interface Omnibox {
  keyword: string;
}

export interface Developer {
  name?: string;
  url?: string;
}

export interface MutualManifest {
  devtoolsPage?: string;
  homepageUrl?: string;
  omnibox?: Omnibox;
}

export interface FirefoxManifest extends MutualManifest {
  developer?: Developer;
}

export interface ChromeManifest extends MutualManifest {
  contentSecurityPolicy?: string;
  versionName?: string;
}

export interface EdgeManifest {}

export interface BrowserManifest extends FirefoxManifest, ChromeManifest, EdgeManifest {}

export interface Manifest extends BrowserManifest {
  name: string;
  author: string;
  version: string;
  defaultLocale: string;
  permissions?: string[];
  description?: string;
  chrome?: ChromeManifest;
  firefox?: FirefoxManifest;
  edge?: EdgeManifest;
}