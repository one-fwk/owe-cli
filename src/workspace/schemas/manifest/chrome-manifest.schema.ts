import { toTypeNumber, toTypeObject, toTypeString } from '../to-schema';

export const chromeManifestSchema = toTypeObject({
  additionalProperties: false,
  properties: {
    contentSecurityPolicy: toTypeString(),
    manifestVersion: toTypeNumber(),
    versionName: toTypeString(),
  },
});