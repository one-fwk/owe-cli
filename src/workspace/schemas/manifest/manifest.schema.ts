import { chromeManifestSchema } from './chrome-manifest.schema';
import { toTypeString, uniqueArrayOfStrings } from '../to-schema';

export const manifestSchema = {
  type: 'object',
  additionalProperties: false,
  properties: {
    name: toTypeString(),
    author: toTypeString(),
    version: toTypeString(),
    description: toTypeString(),
    defaultLocale: toTypeString(),
    // @TODO: Needs fixing
    permissions: uniqueArrayOfStrings(),
    chrome: chromeManifestSchema,
    // edge: {},
    // firefox: {},
  },
  required: [
    'name',
    'author',
    'version',
    // 'description',
  ],
};