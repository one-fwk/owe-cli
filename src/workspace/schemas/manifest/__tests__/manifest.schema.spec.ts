import * as Ajv from 'ajv';

import { manifestSchema } from '../manifest.schema';

describe('manifestSchema', () => {
  let baseSchema: any;
  let validate: Ajv.ValidateFunction;
  let ajv: Ajv.Ajv;

  beforeEach(() => {
    ajv = new Ajv({ useDefaults: true });

    baseSchema = {
      name: '',
      version: '',
      author: '',
    };
  });

  it('should accept required properties without errors', () => {
    expect(ajv.validate(manifestSchema, baseSchema)).toBe(true);
  });

  describe('permissions', () => {
    it('should accept unique array of permissions', () => {
      expect(ajv.validate(manifestSchema, {
        ...baseSchema,
        permissions: [
          '1',
          '2',
        ],
      })).toBe(true);
    });

    it('should NOT have duplicate items', () => {
      expect(ajv.validate(manifestSchema, {
        ...baseSchema,
        permissions: [
          '1',
          '1',
        ],
      })).toBe(false);

      expect(ajv.errors).toMatchSnapshot();
    });

    it('should NOT accept anything but strings', () => {
      expect(ajv.validate(manifestSchema, {
        ...baseSchema,
        permissions: [
          '',
          null,
          1,
          {},
        ],
      })).toBe(false);

      expect(ajv.errors).toMatchSnapshot();
    });
  });
});