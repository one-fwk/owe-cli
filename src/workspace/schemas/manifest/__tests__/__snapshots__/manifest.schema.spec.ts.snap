// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manifestSchema permissions should NOT accept anything but strings 1`] = `
Array [
  Object {
    "dataPath": ".permissions[1]",
    "keyword": "type",
    "message": "should be string",
    "params": Object {
      "type": "string",
    },
    "schemaPath": "#/properties/permissions/items/type",
  },
]
`;

exports[`manifestSchema permissions should NOT have duplicate items 1`] = `
Array [
  Object {
    "dataPath": ".permissions",
    "keyword": "uniqueItems",
    "message": "should NOT have duplicate items (items ## 1 and 0 are identical)",
    "params": Object {
      "i": 0,
      "j": 1,
    },
    "schemaPath": "#/properties/permissions/uniqueItems",
  },
]
`;
